src/include/migraphx/op/slice.hpp

```diff
@@ -90,80 +190,149 @@ struct slice
+    }
 
-        std::vector<std::size_t> new_lens = old_lens;
-        for(std::size_t i = 0; i < axes.size(); i++)
+    /**
+     * Three input version of the normalize_inputs.
+     * This one also checks that the input_axes are valid.
+     */
+    template <class Inds>
+    std::unordered_map<std::string, Inds>
+    normalize_inputs(shape input_shape, Inds input_starts, Inds input_ends, Inds input_axes) const
+    {
+        // normalize the axes
+        auto shape_ndim = input_shape.ndim();
+        std::transform(input_axes.begin(), input_axes.end(), input_axes.begin(), [&](auto axis) {
+            if(axis < 0)
+            {
+                axis += shape_ndim;
+            }
+            if(axis < 0 or axis >= shape_ndim)
+            {
+                MIGRAPHX_THROW("SLICE: entry of input_axes out of bounds");
+            }
+            return axis;
+        });
+        return {{"input_starts", normalize_indices(input_shape, input_starts, input_axes)},
+                {"input_ends", normalize_indices(input_shape, input_ends, input_axes)},
+                {"input_axes", input_axes}};
+    }
+
+    argument compute(const shape& output_shape, std::vector<argument> args) const
+    {
+        auto input       = args[0];
+        auto input_shape = input.get_shape();
+        std::size_t offset = 0;
+        switch(args.size())
         {
-            auto axis            = axes[i];
-            size_t sliced_length = ends[i] - starts[i];
-            // A Numpy indexing convention: a slice size larger than the actual dimension
-            // is legal and the "ends" value is clipped to the axis size
-            new_lens[axis] = std::min(new_lens[axis], sliced_length);
-            if(input_shape.dynamic())
+        case 1: {
+            offset = compute_offset(input_shape);
+            if(output_shape.dynamic())
+            {
+                return {normalize_compute_shape({input_shape}),
+                        [=] { return input.data() + offset; }};
+            }
+            else
             {
-                // TODO: when non-fixed shape slicing is allowed, this will be different than
-                // sliced_length, making use of TBD start/end values.
-                std::size_t sliced_min_length = ends[i] - starts[i];
-                // if the slice size is smaller than maxes but larger than mins
-                new_mins[axis] = std::min(sliced_min_length, new_mins[axis]);
+                return {output_shape, [=] { return input.data() + offset; }};
             }
         }
-        if(input_shape.dynamic())
-        {
-            return shape{t, new_mins, new_lens, {}};
+        case 3: {
+            shape calc_shape;
+            visit_all(args[1], args[2])([&](auto input_starts, auto input_ends) {
+                auto norm_inputs = normalize_inputs(input_shape, input_starts, input_ends);
+                offset = compute_offset(input_shape, norm_inputs.at("input_starts"), this->axes);
```

Comment:
`std::size_t offset = compute_offset(input_shape, norm_inputs.at("input_starts"), this->axes);`